### PR TITLE
[hail] Inline ApplyIR in compile lowering pass.

### PR DESCRIPTION
**Pull request by tpoterba**
We're currently emitting the explicit node (without optimization!).
This design is incrementally better, and lets us do ptyping more easily.

The right solution is to do generate a method as the node suggests, but
there are some issues to sort out here, like how to return a missing
value. We may need to return a (possibly null) pointer to an allocated
value, which could be inefficient. Pushing ptypes/requiredness fully
through the system would let us avoid this in many cases.

Stacked on #8084, don't review until that goes in.

### CODE REVIEW

File: hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala [hunk 1]
```diff
@@ -328,7 +328,7 @@ class FunctionBuilder[F >: Null](val parameterTypeInfo: Array[MaybeGenericTypeIn
       cn.accept(cw)
       bytes = cw.toByteArray
 //       This next line should always be commented out!
-//      CheckClassAdapter.verify(new ClassReader(bytes), false, new PrintWriter(sw1))
+      CheckClassAdapter.verify(new ClassReader(bytes), false, new PrintWriter(sw1))
```
Comment (akotlar) [inline hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: This next line should be commented out!

Comment (akotlar) [reply hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: This needs a rebase I think. Not sure why the diff hasn’t included the changes from the stacked pr

Comment (akotlar) [reply hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: Ohh, probably the stacked label blocks.

Comment (akotlar) [reply hail/src/main/scala/is/hail/asm4s/FunctionBuilder.scala]: But the above comment should still be removed?

**Comment by tpoterba** [comment]
oops, already have one open

**Comment by tpoterba** [comment]
sorry, review the other one! just pushed a commit to retest (this one was passing)

**Pull request closed without merging**